ref(server): Manually control concurrent upstream requests

RaduW opened the pull request:
WIP Move implementation from all upstream requests being sent when they are created
to sending them only when there are free http connections available.

RaduW committed:
before removing the transformer from SendRequest

RaduW committed:
removed ResponseTransformer

RaduW committed:
WIP push requests into the queue (no priority yet)

Still to do:
- consuming loop not done
- figure out if we can do consuming loop somehow in the actor thread or
we need to spanw thread ( and change request sender not to receive self)

RaduW committed:
WIP added send message loop.

- still missing free http connection counters

RaduW committed:
WIP Before switching to hack week project

RaduW committed:
merged master

RaduW committed:
split messages into high/low priority queues

RaduW committed:
removed &self from WithRequestPriority::priority() signature

priority is type specific, not instance specific and therefore should
not have &self in the signature of WithRequestPriority::priority()

RaduW committed:
now tracking request finishing and queue size

RaduW committed:
Added some docs

RaduW committed:
Changed TrackedFuture - instead of sending unit now sends a unit struct.

Changed for interface legibility of.

RaduW committed:
Finished implementation, after talking with @untitaker

RaduW committed:
fixed oneshot::channel communication

RaduW committed:
fixed integration tests

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -133,6 +157,47 @@ impl UpstreamRateLimits {
+/// An type implementing this trait can be placed in a queue according to its priority.
+/// This trait is implemented by message types that need to be sent via http
+pub trait WithRequestPriority {
+    fn priority() -> RequestPriority;
```
I think we can move this function into `UpstreamQuery`, as this is the only place where this functionality is needed. That saves us from implementing a separate trait. 

Also: Consider to add a default implementation that returns `RequestPriority::Low`. This is very explicit in places where we need high prio.

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -16,13 +35,15 @@ use serde::de::DeserializeOwned;
-use crate::utils::{self, ApiErrorResponse};
+use crate::metrics::RelayHistograms;
+use crate::utils::{self, ApiErrorResponse, IntoTracked, TrackedFutureFinished};
+use futures::sync::{mpsc, oneshot};
```
_nit_: Move the `futures` import two groups up to external dependency imports.

RaduW replied:
yes

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -133,6 +157,47 @@ impl UpstreamRateLimits {
+    max_inflight_requests: usize,
+    num_inflight_requests: usize,
+    // high priority messages
+    hp_messages: VecDeque<UpstreamRequest>,
```
_nit_: Rename to `high_prio_requests` for clarity. Same below for low prio.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -181,18 +245,22 @@ fn handle_response(
             config,
             auth_state: AuthState::Unknown,
+            // TODO get the real value from a config and use it with ClientConnector::limit
+            max_inflight_requests: 100,
```
Please add a config value to `config.limits`.

RaduW replied:
Yes, after looking a bit more I realised that we already have a configuration variable for this `Limits::max_concurent_requests` which is already used to configure the `ClientConnector::limit()`.

Changed from hardcoded value to `Config::max_concurent_requests()`

jan-auer replied:
That's perfect, of course that's the one we want anyway.

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -204,16 +272,14 @@ impl UpstreamRelay {
-        F: FnOnce(&mut ClientRequestBuilder) -> Result<ClientRequest, ActixError>,
-        P: AsRef<str>,
-    {
+    fn send_http_request(&mut self, request: UpstreamRequest) -> ResponseFuture<(), ()> {
```
_nit_: Rename this to `send_request` and the other function to `enqueue_request` for clarity.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -246,21 +325,71 @@ impl UpstreamRelay {
-            .conn_timeout(self.config.http_connection_timeout())
             // This is the timeout after wait + connect.
             .timeout(self.config.http_timeout())
+            .conn_timeout(self.config.http_connection_timeout())
```
_nit_: Move this to the previous location (under `wait_timeout`) again to reduce the diff.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -246,21 +325,71 @@ impl UpstreamRelay {
+            .and_then(handle_response)
+            .then(|x| {
+                response_sender.send(x).ok();
+                futures::future::ok(())
```
_nit_: You can simplify this by returning `Ok(())`, which will be automatically transformed into a future.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -246,21 +325,71 @@ impl UpstreamRelay {
+                self.lp_messages.push_front(request);
+                metric!(
+                    histogram(RelayHistograms::UpstreamLowPriorityMessageQueueSize) =
+                        self.lp_messages.len() as u64
```
Please use a single metric here and tag the queue priority. This will allow you to deduplicate the code like this (pseudo-code):

```rust
let queue = match priority {
  RequestPriority::Low => &mut self.lp_messages,
  RequestPriority::High => &mut self.hp_messages,
};

queue.push_front(request);
metric!(histogram(RelayHistograms::UpstreamQueueSize = ..., priority = priority));
```

RaduW replied:
done, didn't manage to use directly `priority` in the histogram, I'm using a `fn name(&self)` method i.e. `priority.name()`

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -462,55 +612,83 @@ impl SendRequest {
+/// the mpsc channel or this handler, it would have not dealt with dropped futures.
+/// Weather the added complexity of this design is justified by being able to handle dropped
+/// futures is not clear to me (RaduW) at this moment.
+impl StreamHandler<TrackedFutureFinished, ()> for UpstreamRelay {
```
This is up for discussion: I think it would be simpler to implement this as a plain message with `context.notify()` for the following reasons:
1. you don't need to handle and worry about stream errors
2. you don't need to keep the stream around in the upstream actor struct
3. it aligns very well with other handlers in the upstream actor

RaduW replied:
done

jan-auer commented on relay-server/src/utils/tracked_future.rs:
```diff
@@ -0,0 +1,71 @@
+where
+    F: Future,
+{
+    fn to_tracked(self, notifier: Sender<TrackedFutureFinished>) -> TrackedFuture<F>;
```
_nit_: Usually these methods would be called with a verb like `.track()`. I'd propose to rename.

RaduW replied:
ok

jan-auer commented on relay-server/src/utils/tracked_future.rs:
```diff
@@ -0,0 +1,71 @@
+                    .map_err(|_| {
+                        log::error!("TrackedFuture could not notify completion");
+                    })
+                    .ok();
```
_nit_: Deduplicate in a function:

```rust
fn notify(&mut self) {
    if self.notified { return; }

    self.notified = true;
    self.notifier
        .try_send(TrackedFutureFinished)
        .map_err(|_| {
            log::error!("TrackedFuture could not notify completion");
        })
        .ok();
}
```

Also: Please add a comment explaining why you eagerly run this here and not only in drop.

RaduW replied:
good point.

jan-auer reviewed (changes requested):
Thanks, the approach looks great. Below are nit picks and naming changes.

RaduW committed:
Merged UpstreamQuery and WithRequestPriority traits

RaduW committed:
mostly renames

RaduW committed:
all review comments except switching from channel to messages

RaduW committed:
Moved from mpsc channels in tracked_future.rs to Actix::Recipient

Instead of using an mpsc channel to notify the end of a tracked future
now we use an `Recipient<TrackedFutureFinished>` with the handler
implemented in `UpstreamRelay`

RaduW committed:
Merge branch 'master' into ref/priority-upstream

RaduW committed:
Updated changelog.
